src/taskgraph/util/cached_tasks.py

```diff
@@ -53,31 +56,43 @@ def add_optimization(
 
     # We'll try to find a cached version of the toolchain at levels above and
     # including the current level, starting at the highest level.
-    # Chain-of-trust doesn't handle tasks not built on the tip of a
-    # pull-request, so don't look for level-1 tasks if building a pull-request.
     index_routes = []
     min_level = int(config.params["level"])
-    if config.params["tasks_for"] == "github-pull-request":
-        min_level = max(min_level, 3)
     for level in reversed(range(min_level, 4)):
         subs["level"] = level
         index_routes.append(TARGET_CACHE_INDEX.format(**subs))
 
-        taskdesc["optimization"] = {"index-search": index_routes}
+    # Pull requests use a different target cache index route. This way we can
+    # be confident they won't be used by anything other than other pull
+    # requests.
+    if config.params["tasks_for"].startswith("github-pull-request"):
+        # Check the PR bucket first.
+        subs["head_ref"] = config.params["head_ref"]
+        if subs["head_ref"].startswith("refs/heads/"):
+            subs["head_ref"] = subs["head_ref"][11:]
+        index_routes.insert(0, TARGET_PR_CACHE_INDEX.format(**subs))
```

Comment:
Hm, maybe we should check for the cached task in the PR namespace last?

Reply:
I think that would be more consistent with what we do now: we look at the most trustworthy source first. In reality though, I don't see why it would make a difference unless someone finds hash collisions, or a cache digest is seriously misconfigured.